### ISSUE TITLE
[css-color-5] Make channel values in relative color syntax optional

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#10689]
To reduce complexity for adding transparency to a color (see #10678), the channel values should be made optional, as suggested by @romainmenke.

So, the syntax of the `rgb()` function as an example would then be

```ebnf
<modern-rgb-syntax> = rgb( [ from <color> ]?
        [ <number> | <percentage> | none]{3}?
        [ / [<alpha-value> | none] ]?  )
```

allowing to define a semi-transparent black as `rgb(from black / 0.5)`.

The syntax above also allows to define `rgb(from black)`, which might not be useful by itself. If we decide to enforce one of the components to be there, the syntax may then look like

```ebnf
<modern-rgb-syntax> = rgb( [ from <color> ]?
        [ [ <number> | <percentage> | none]{3}?
        [ / [<alpha-value> | none] ]? ]!  )
```

Sebastian

**Comment by romainmenke** [comment]
`rgb(from black)` would also be useful when gamut mapping is limited to certain notations.

Currently there is a resolution to only have gamut mapping when the color was described with some functions, like `oklch()`.

- `color(display-p3 1.2 -0.1 05)` -> is **not** gamut mapped
- `oklch(from color(display-p3 1.2 -0.1 05))` -> is gamut mapped

**Comment by tabatkins** [comment]
I don't think the difference between `oklch(from color(display-p3 1.2 -0.1 05) l c h)` and `oklch(from color(display-p3 1.2 -0.1 05))` is meaningful. ^_^

For the original case, while omitting the non-alpha components is grammatically distinguishable, I don't think it's a great idea for readability. You can't omit channels in the normal function version, which would mean the relative syntax diverges from the absolute syntax in more ways; currently it's just an addition (letting you use channel keywords).

Basically, I just don't think having to write `r g b` in `rgb(from var(--foo) r g b / .5)` is much of an imposition. If it really is, then having a more dedicated function, like `opacity(var(--foo) .5)` would be *even shorter and clearer*.

**Comment by francescorn** [comment]
Thank you for opening @SebastianZ.

@tabatkins A function such as `opacity` would be nice where you can input a function and control specific opacities such as `opacity((element/variable) value)`. @tabatkins Then similar to what was brought up in: https://github.com/w3c/csswg-drafts/issues/10677 and https://github.com/w3c/csswg-drafts/issues/10678, you can control the specific opacity for many elements and variables. 

@romainmenke Maybe we can have functions to opt-in or opt-out gamma mapping? I see a javascript example in [css-color-4](https://www.w3.org/TR/css-color-4/), maybe we can automate it in CSS? Such as `gamma-on(color(display-p3 1.2 -0.1 05)`, and `gamma-off(color(display-p3 1.2 -0.1 05)`.

**Comment by romainmenke** [comment]
My suggestion was to make use of what already exists and to make this (marginally) easier to type for CSS authors.

My thinking was that the effort to make the channel keywords optional would be a less than introducing new color related functions.

Having sane default behavior even when arguments are omitted is also often a good thing for CSS authors. (e.g. recent changes `clamp()`)

I agree that the added value of saving a few keystrokes is slim here, but I think it is balanced by the (seemingly?) low effort to make this change and how common it is to only want to adjust the opacity of a color.

I don't have a strong opinion on this, I think that `rgb(from pink / 0.3)` is kinda nice, I've often found myself only changing opacity and wanting such a simplified syntax. But I never wanted it that much that I opened an issue about it myself :)

> I just don't think having to write `r g b` in `rgb(from var(--foo) r g b / .5)` is much of an imposition.

I agree, but at the same time I also each time think: "why do I need to?"

**Comment by francescorn** [comment]
I am okay with a rgb(from pink / 0.3) syntax.

**Comment by romainmenke** [comment]
This is also something that authors are already trying to do today, although with a slightly different syntax: https://github.com/stylelint/stylelint/issues/8302

```
rgba(#ff611d, 0.5)
```

This author is trying to change only the alpha of a given color.
I suspect this is something that does work in Sass?

**Comment by jamiebuilds** [comment]
I was just about to create an issue requesting this syntax when I found this one.

I was surprised to find that the `color(from black / 15%)` didn't _already_ work. I thought it would be a useful way to be indifferent to color spaces being specified elsewhere.

It would be great if things like this worked:

```css
.lightbox::backdrop {
  background: color(from light-dark(white, black) / 85%);
}

/* or */

@keyframes glow-pulse {
  0%, 100% { box-shadow: 0 0 10px color(from var(--glow-pulse-color) / 25%); }
  50%      { box-shadow: 0 0 10px color(from var(--glow-pulse-color) / 45%); }
}
```

Particular for the case of `var()` I would hope that it would preserve the color space I'm using<sup>[1]</sup>:

```css
.emerald-city-welcome-sign {
  animation: 1s ease glow-pulse infinite alternate;

  --glow-pulse-color: rgb(0% 100% 0%);

  /* Make it as green as the device supports */
  @media (color-gamut: p3) {
    --glowing-color: color(display-p3 0 1 0);
  }
}
```

> [1] Note: It [appears](https://codepen.io/pen/?editors=1100) that using `color(from var(--x) srgb r g b / 50%)` doesn't "clamp" a P3 color to the sRGB color space so this last point might just be redundant)

**Comment by svgeesus** [comment]
[poll concluded](https://mastodon.scot/@svgeesus/114574983650035139) with 152 responses. 

- **55%** preferred `alpha(var(--mycolor) 0.6)`
- **42%** preferred `opacity(var(--mycolor) 0.6)`
- **3%** preferred some other option.

![screenshot of the poll results, on Mastodon](https://github.com/user-attachments/assets/dfd9e3d9-2d20-4707-8dd0-1414f4ddb55e)

Several people commented that they would expect `opacity()` to affect images and child elements, in other words like the `opacity` property.

The polled syntax set the alpha to a fixed value. Relative values would also be possible, with an `alpha` keyword, and calc:

`alpha(var(--mycolor) calc(alpha * 0.8))`

In general, having a dedicated function to adjust alpha values of colors seemed to be liked compared to the convoluted paths that people use now (including having to set the colorspace, which makes no difference to alpha).

So - while the question of making channel values optional is still open, the majority use case seems to be addressed by a new, dedicated "alpha RCS"

**Comment by romainmenke** [comment]
> Several people commented that they would expect `opacity()` to affect images and child elements, in other words like the `opacity` property.

I voted for `opacity` but changed my mind after reading those comments and now also favor `alpha()`.

**Comment by svgeesus** [comment]
> I voted for `opacity` but changed my mind after reading those comments and now also favor `alpha()`.

I also initially preferred `opacity()` and now prefer `alpha()`

**Comment by svgeesus** [comment]
Agenda+ to discuss the proposal to add an alpha-only RCS function `alpha()` which has no colorspace declaration and affects only the alpha, returning a color with adjusted alpha in whatever colorspace it was originally using.

**Comment by svgeesus** [comment]
The CSS Working Group just [discussed](https://www.w3.org/2025/08/19-css-color-minutes.html#c3a6) `[css-color-5] Make channel values in relative color syntax optional` and agreed to the following:

 * `RESOLUTION: add an alpha-only RCS function alpha()`

<details><summary>The full IRC log of that discussion</summary>
ChrisL: the idea is lot of use case are deal by a new relative color syntax when you dont specifify the color space because you just want to change the alpha <br>
<TabAtkins> `alpha(&lt;color> / &lt;alpha>)` <br>
ChrisL: Im proposing a new relative syntax for just changing the alpha : alpha(&lt;color> / &lt;alpha>) <br>
ChrisL: 2 syntaxes proposed : alpha or opacity <br>
ChrisL: people prefers alpha <br>
TabAtkins: I think its a good move <br>
<ChrisL> proposal: add an alpha-only RCS function `alpha()` <br>
lea: I though the thread was about making components optional in RCS <br>
ChrisL: the good part of this new syntax is that you dont have to specifiy the color space which prevent mistakes <br>
`RESOLUTION: add an alpha-only RCS function `alpha()`` <br>
</details>

**Comment by svgeesus** [comment]
We [currently have]():

> If the alpha value of the relative color is omitted, it defaults to that of the [origin color](https://drafts.csswg.org/css-color-5/#origin-color) (rather than defaulting to 100%, as it does in the absolute syntax).

One of those annoying edge cases that still has to be covered: the pointless "do nothing" syntax:

```css
alpha(from var(--mycolor))
```

Options:

- don't allow this, it is pointless. Require alpha always for this form. But this is different to all the rest of RCS
- allow it, for consistency, it is valid and has an obvious interpretation and the alpha is that of the origin color

**Comment by svgeesus** [comment]
My preference is to allow it, we already allow other no-op syntax

```css
--foo: oklch(90% 0.06 83);
--bar: oklch(from var(--foo) l c h);
```

**Comment by svgeesus** [comment]
Because this RCS is not extending an existing color function, I made the `from color` mandatory not optional. Otherwise there is no defined result.

**Comment by Crissov** [comment]
> The syntax above also allows to define `rgb(from black)`, which might not be useful by itself.

`rgb(from currentcolor)` or `rgb(from accentcolor)` might, though. (Alas, I cannot think of a scenario where it does.)

**Comment by svgeesus** [comment]
We have [this example](https://drafts.csswg.org/css-color-5/#ex-alpha-replace) and [this other example](https://drafts.csswg.org/css-color-5/#ex-alpha-change)

**Comment by svgeesus** [comment]
Three WPT have been added

- [alpha-color-computed.html](https://wpt.fyi/results/css/css-color/parsing/alpha-color-computed.html)
- [alpha-color-parsing-invalid.html](https://wpt.fyi/results/css/css-color/parsing/alpha-color-parsing-invalid.html)
- [alpha-color-parsing-valid.html](https://wpt.fyi/results/css/css-color/parsing/alpha-color-parsing-valid.html)

**Comment by Crissov** [comment]
It is consistent to require `from` in there, but it would be convenient if it was optional.

**Comment by yisibl** [comment]
> However, to me as a programmer, a function name should convey its meaning. And to me any "alpha(..)" function would mean a function *returning* an alpha value (so here extracting the color alpha channel value). Which would be a cool function to have, but is not the intent here.
> An allowed simplified use of color-mix(color %) would make more sense to me.
> https://indieweb.social/@villapirorum/114575974057220952

@svgeesus What are your thoughts on this simplified version of `color-mix()`? Has the CSS Working Group discussed it?
